core/src/main/java/org/htsjdk/core/utils/PathSpecifier.java

```diff
@@ -0,0 +1,238 @@
+ *       hier_part     = ( net_path | abs_path ) [ "?" query ]
+ *       net_path      = "//" authority [ abs_path ]
+ *       abs_path      = "/"  path_segments
+ *
+ * A URI is absolute if, and only if, it has a scheme component.
+ * A URI is opaque if, and only if, it is absolute (has a scheme) and its
+ * scheme-specific part does not begin with a slash character ('/')
+ *
+ * A relative reference that does not begin with a scheme name or a
+ * slash character is termed a relative-path reference.
+ *
+ * A relative reference beginning with a single slash character is
+ * termed an absolute-path reference, as defined by <abs_path> in
+ * Section 3.
+ *
+ * URI that do not make use of the slash "/" character for separating
+ * hierarchical components are considered opaque by the generic URI
+ * parser.
+ */
+public class PathSpecifier implements PathURI, Serializable {
+    private static final long serialVersionUID = 1L;
+
+    private final String    rawInputString;     // raw input string provided by th user; may or may not have a scheme
+    private final URI uri;                      // working URI; always has a scheme (assume "file" if not provided)
+    private String          pathFailureReason;  // cache the reason for "toPath" conversion failure
+
+    private transient Path cachedPath;         // cache the Path associated with this URI if its "Path-able"
+
+    public PathSpecifier(final String rawInputString) {
+        ParamUtils.nonNull(rawInputString);
+        this.rawInputString = rawInputString;
+
+        URI tempURI;
+        try {
+            // If the input URI already has a scheme (including a "file" scheme), we assume its already properly
+            // escaped. If no scheme component is present, then assume its a raw path on the local file system, so
+            // try to get a Path first, and then recreate the URI by retrieving it from the resulting Path. This
+            // ensures that input strings that contain embedded characters that would otherwise be interpreted as
+            // URI syntax (such as embedded fragment specifiers ("#") that are valid in file names) are properly
+            // escaped.
+            tempURI = new URI(rawInputString);
+            if (!tempURI.isAbsolute()) {
+                // NOTE: This case (no scheme) is the only one where we resolve the URI to a Path at construction time.
+                try {
+                    setCachedPath(Paths.get(rawInputString));
+                    tempURI = getCachedPath().toUri();
+                } catch (InvalidPathException e) {
+                    throw new IllegalArgumentException(e.getMessage(), e);
+                }
+            }
+            uri = tempURI;
+            if (!uri.isAbsolute()) {
+                // assert the invariant that every URI we create has a scheme, even if the raw input string does not
+                throw new HtsjdkException("URI has no scheme");
```

Comment:
We should have either more specific exceptions (never throw a raw one) or in this case it should be an `IllegalArgumentException` as it is a problem with the provided param.

Reply:
Changed this one to` IllegalArgumentException`. All of the others are HtsjdkIOException (subclass of `HtsjdkException`).

If `HtsjdkException` is truly intended to only be a base class, it should probably have protected constructors.